pep-9999.rst

```diff
@@ -0,0 +1,464 @@
+fixed shape, fewer barriers, inlining, etc.). Freezing objects can
+allow Python to implement more efficient caching mechanisms and
+enable compiler optimisations that rely on immutability
+assumptions.
+
+
+Note that none of the above are *guaranteed* with immortal
+objects as these may still change if a program manipulates an
+object without knowing that it is immortal.
+
+
+Specification
+=============
+
+Changes to Python objects
+-------------------------
+
+Every Python object will have a flag that keeps track of its
+immutability status. Details about the default value of
+this flag is discussed further down in this document. 
+
+The flag can be added without extending the size of the
+Python object header. We initially intend for this flag 
+to be implemented by bit-stealing from the reference 
+counter, but this will change in our planned extensions.
+
+
+Implementation of Immutability
+------------------------------
+
+Immutability is enforced through run-time checking. The macro
+``Py_CHECKWRITE(op)`` is inserted on all paths that are guaranteed
+to end up in a write to ``op``. The macro inspects the immutability
+flag in the header of ``op`` and signals an error if the immutability
+flag is set.
+
+A typical use of this check looks like this:
+
+.. code-block:: c
+
+	if (!Py_CHECKWRITE(op)) {        // perform the check
+	    PyErr_WriteToImmutable(op);  // raise the error if the check fails
+	    return NULL;                 // abort the write
+	}  
+	... // code that performs the write
+
+
+As writes are common but lack a common path that most writes to through
+the PEP requires a ``Py_CHECKWRITE`` call, there are several places in
+the CPython code base that are changed as a consequence of this PEP.
+So far we have identified around 70 writes spread across a dozen files.
+
+
+New Obligations on C Extensions
+-------------------------------
+
+Because our implementation builds on information available to the CPython
+cycle detector, types defined through C code will support immutability 
+"out of the box".
+
+C extensions that perform writes to data which can be made immutable
+should add the ``Py_CHECKWRITE`` macro shown above on all paths in the
+code that lead to writes to that data.
+
+**TODO**
+  If we don't have a way to tell whether or not an extension 
+  supports immutability, then freeze might end up "leaking"
+  mutable data. Let's discuss this.
+
+We are discussing adding a ``__freeze__`` hook that will be called 
+after an object has been made immutable. This hook could then be used 
+to freeze or otherwise manage any other state on the side that is
+introduced through a C-extension.
+
+
+Freezing Type Which are Not Immutability-Aware
+----------------------------------------------
+
+**TODO** Decide what happens when we attempt to freeze an
+object and discover an object which is not immutability-aware.
+
+1. Throw an exception
+2. Leave it and live with the unsoundness
```

Comment:
My vote is for this. Not sure what is meant by "immutability-aware" but our SOP is that we use `tp_traverse` to freeze things and if someone wants to participate in the freezing process they should register with that, which they should really be doing already (except for hidden c state as mentioned above)

Reply:
Great -- let's talk about this in a call next week!

Reply:
But I don't think you are right (but maybe you are). I understand we can find data and flip a bit in it, but if the C extension implements its own writes to that data, it needs to look at that bit too.

Reply:
Essentially the only things that can be made immutable are Python objects. If the extension author has python objects in an object's graph that need to be made immutable, they should already be registering them with `tp_traverse`. If there is C data living in their custom object's data region, then need to use `Py_CHECKWRITE` to check the object's immutability status and then act accordingly.